### PR TITLE
Fix/docstrings

### PR DESCRIPTION
**Pull request by abegong**
I spent my flight to Dublin cranking away at documentation.

It's coming together pretty nicely.

Can you take a look at the first five expectations and see if you like this format? If so, we can roll it out to the rest of them.

Also, please take a look at the diffs in the rest of the docs. I've moved some pieces around and fixed some lingering inaccuracies.

Sample pics for expectation docs are below:

screen shot 2017-12-11 at 6 15 26 am
screen shot 2017-12-11 at 6 15 34 am
screen shot 2017-12-11 at 6 15 44 am
screen shot 2017-12-11 at 6 15 57 am

### CODE REVIEW

**Comment by jpcampb2** [comment]
Holy millions of commits! This looks amazing.

Lots of notes:

1. We currently have a secret parameter called `meta_notes`, which should be added to all expectations or removed (or at least acknowledged). I'm okay to just remove it, but you'd flagged it as desirable so want to get your sense.
2. We don't currently enforce results being json-serializable (but could in the expectation wrappers; this needs a new issue perhaps?)
3. I think docstrings should describe true_value value or other expectation-specific outputs...how would that work with this format? (e.g. `observed_partition`?)
4. we should change `to_not_be_null` to `to_be_not_null`
6. I think there was back-and-forth on inclusion or not of `expect_column_values_to_be_unique` (vs `expect_column_proportion_of_unique_values_to_be_between` which includes it. I think having both is fine, though it marginally increases the implementation burden for a new DataSet. Flagging.
6. In an effort to improve searching, I think it will be useful to ensure that we use all of the terms in the expectation's name in the docstring. So `expect_row_count_to_be_between` should have a docstring initial line like: "Expect the number of rows (row count) to be between two values"
7. I do see a few very small typos. I can't modify the PR since it's an external branch, so I'm just going to approve and then we can fix those as we continue to update the docs.
8. I don't think we want to make sphinxcontrib a requirement for any user of great expectations. That's just for building the docs, right?
9. There are documentation errors in some of the support functions for distributional expectations, but they are not the result of formatting errors, so should not hold up this PR.

**Comment by abegong** [comment]
1. I think we should remove `meta.notes` in the sense of not giving it special treatment. Users are still free to populate the field if they want.

I know that's different from what I said before, and I could be persuaded to change my mind back. I just want to be careful about not packing too many random parameters onto expectations.

2. Check out PR #156. It includes a general serialization check.

3. Agreed. Does this example look okay to you?

screen shot 2017-12-12 at 2 33 36 am


4. Done.

5. Pushback: they behave differently with mostly, so they deserve to remain separate.

6. +1. Can I propose that you screen all expectations using that criteria as part of the PR for (7)?

7. +1 Sounds good.

8. I'm not sure how that works. You need to have it installed to build the docs locally. And I think it needs to be in requirements.txt for readthedocs to do its thing.

In other words, I think it might be a required dependency.

9. +1. Put this in the PR for (7) as well?

-----

Next steps:

If I read this right, we're proposing that
* 1 and 3 go in this PR
* 6, 7, and 9 go in a separate PR.
* 2 and 4 are already resolved.
* 5 and 8 are good as they stand.

Also, I'm going to continue converting all the expectation docstrings to the new format. We're only a few hours away from getting them all done, and I'd prefer to get everything to a stable place.

Is that right?

**Comment by jpcampb2** [comment]
1. Agreed. Meta_notes should not be a thing.
2. I can't see a general serialization check on the output of expectations in #156. To clarify, I meant the expectation output, vs. the expectation config (these docs promise that is also serializable)
3. Rather than putting the expectation-specific return values into a Notes section at the end, can they be embedded into the return documentation directly (so indented and not below "notes")
4. Done.
5. I'm okay with them remaining separate. Re: that they do different things: isn't `expect_column_proportion_of_unique_values_to_be_between(column, 0.9, 1)` the same as `expect_column_values_to_be_unique(column, mostly=0.9)` (and similarly for other parameter values?) (this is an aside, no problem for me to continue with both at this point since we're not thinking deeply about the grammar of expectations at this point).
6. Agreed
7. Agreed
8. I still think it's incorrect to declare a doc building tool as a dependency of the package, and we ought to figure out how to separate that. I think there are a few avenues, one of which is to have a separate requirements for documentation (e.g. requirements-docs.txt, which we declare to readthedocs). Another seems to be to add a "development requirements" section (see https://stackoverflow.com/questions/28509965/setuptools-development-requirements). Either way my concern is that we not push to pypi with ephemeral dependencies.
9. Agreed.

So in summary: I'm still holding out for a small reversion on 8, confirmation of change on 1, clarification on 2, and it seems like the other things are in works or for separate subsequent PRs.

Great job on this!

**Comment by abegong** [comment]
1. Agreed. I said earlier that this should go in this PR. Can we put it in the next one?
2. You're right. The serialization chcek in the other branch is for expectation configs, not expectation results. Can we also put this in the next PR?
3. I tried that first, and the formatting isn't as clean (see below). I tried a bunch of options and the one I posted earlier was the best I could find.
8. I get it. If I take out those packages from requirements.txt, will figure out how to keep RTD happy?

screen shot 2017-12-13 at 5 33 03 am

**Comment by jpcampb2** [comment]
Pushing things to another PR is fine for me given that this is so big already and focused on documentation only. I've created issues for (1 #157) and (2 #158) above.

For formatting: understood. How the return block is amended to read:
"Exact fields vary.... See Notes section for a description of expectation-specific fields in the return object." (That can also wait for a new PR since this one was mostly about formatting; flagged it in #159).

I also created an issue for requirements #160

**Review by jpcampb2 — APPROVED** [review]
Excited for this update.
